text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,267 @@
+  end
+  ```
+  
+  Currently and counter the natural expectation, this code hangs since the Range is *infinite* when the file f existed but was empty. The reason is that in this case the Range expands to `Range[USize](1, 0, 1)` which again in Python and Rust would be considered empty but is currently infinite in Pony. While there are legitimate cases in which the current `Range` implementation returns infinite iterators, this particular case seems illogical since the iterator should produce a number of elements of `step` spacing on the *trajectory from min to max*. In the example above, no such elements can exists due to the sign mismatch between min, max, and step, and the returned Range should be empty rather than infinite -- like it is in Rust, Python, and others. The only reason why Pony creates an *infinite* range in this situation is that it, algorithmically, seems to be saying: I know there are no elements (the implementation actually checks for this "no progress possible" situation), but I will walk indefinitely in the wrong direction because you told me so and hence I am infinite, even though the number of elements on the expected tranjectory is zero. The big issue with this, beside the creation of unwanted infinite loops, is that it also precludes or makes very cumbersome the expressive use of the bounds like in the csv file example above in which the programmer *knows* and *makes use of* the fact that if the bounds (and the step) are not compatible sign-wise, the range will have zero iterations, i.e. be *empty*.
+
+  The more legitimate classification of Ranges as infinite, for example in cases when `min` and `max` are different, finite, and `step == 0`, are not changed in this RFC. For such cases, at least the argument could be made that the number of iterations should become infinitely large. For these, Pony currently provides a test with `is_infinite(): Bool`. Since currently no empty Ranges are recognized, there is also no test for the far more likely to occur empty cases. For example, Rust provides a `is_empty()` method for these cases and this RFC proposes to add `.is_empty(): Bool` to Pony that behaves like the following Rust example:
+  
+  ```rust
+  assert!( (3..3).is_empty());
+  assert!( (3..2).is_empty());
+  ```
+  
+  While one could maybe argue that a `for` loop could be bracketed by a test for `is_infite()` in the above example,
+  
+  ```pony
+  ...
+  let f: Array[Array[String]] = csv.read("data.csv", ",")
+  let lines = Range(1, f.size()) 
+  if not lines.is_infinite() then
+    for idx in lines do // start at 1 to skip header line
+      ... // perform only if f is not empty and has at least 1 data line
+    end
+  end
+  ```
+  
+  it is fair to mention that currently no `Pony`-labeled code on Github other than test code for Range exists that would make use of .is_infinite(). Also, the above is unnecessarily cumbersome, and again prevents the expressive use of the Range bounds with calculated values.
+
```

Comment:
I think it would help to show before and after code and then explain why you think the after is more expressive. A statement like "more expressive" doesn't really mean much. Everyone has their own idea of what "expressive" might mean.